build: macOS toolchain simplification and bump

dongcarl opened the pull request:
This PR achieves 3 main things:
1. It simplifies the macOS SDK generation by putting the logic inside a (semi-)portable python3 script `gen-sdk`
2. It transitions us to using `libc++` headers extracted from the `Xcode.app`, which is more correct as those headers better match the `.tbd` library stubs we use from the `MacOSX.sdk` (located under the same `Xcode.app`). Previously, we used `libc++` headers copied from our downloaded, pinned clang (see `native_cctools.mk`).
3. It bumps the macOS toolchain in a way that fulfills all of the following constraints:
    1. The new SDK should support compiling with C++17 (our current one doesn't)
    2. The new toolchain should not change our minimum supported macOS version (`-mmacosx-version-min`)
    3. The new toolchain should expect to use a version of `cctools` that is supported by https://github.com/tpoechtrager/cctools-port

For the constraints in (3), you can reference [this chart](https://en.wikipedia.org/wiki/Xcode#Xcode_7.0_-_11.x_(since_Free_On-Device_Development)) to see that the newest toolchain we can use with our `cctools-port` is `11.3.1`, and the rest of the constraints were tested with local builds.

#### But [the other Wikipedia chart](https://en.wikipedia.org/wiki/Xcode#Xcode_11.x_(since_SwiftUI_framework)) says that the "min macOS to run" for Xcode 11.3.1 is 10.14.4, doesn't that violate constraint (ii)?

This confused me at first too, but the "min macOS to run" is for the Xcode.app App itself. The SDK still supports 10.12, as evident in a few plist files and as proven through local builds.

#### Why bundle all of this together in a single PR?

We need (1) and (2) together, because if we don't, manually adding the `libc++` headers and writing that out in a `README.md` is going to result in a lot of user error, so it's great to have these together to be more correct and also make it easier on the user at the same time.

We need (3) together with everything else because bumping (or in the case of (1), renaming) the SDK requires some human coordination and may break some builds. And since it's not that complicated a change, it makes sense to do it together with the rest.

dongcarl committed:
contrib: macdeploy: Correctly generate macOS SDK

Previously, we did not include the macOS SDK libc++ headers in our SDK
creation process and instead used whichever libc++ headers shipped with
the clang package we downloaded in depends.

This change adds a script (which works on both GNU/Linux and macOS) to
correctly generate the macOS SDK including the libc++ headers. This can
be thought of as a simplified rewrite of tpoechtrager's script:

https://github.com/tpoechtrager/osxcross/blob/d3392f4eae78f3fa3f1fd065fa423f2712825102/tools/gen_sdk_package.sh

The location within the SDK where we place the libc++ headers is chosen
such that clang's search path detection logic for sysroots would pick up
the headers properly.

We also document this change.

dongcarl committed:
native_cctools: Don't use libc++ from pinned clang

Now that we include the macOS SDK libc++ headers in our macOS SDK
tarball, we no longer need this hack to use the libc++ from our pinned
clang.

dongcarl committed:
contrib: macdeploy: Use apple-sdk-tools instead of xar+pbzx

laanwj commented:
> For anyone who's looked at the last PR, this one just has a simpler sdk-gen.sh script that I wrote myself, and includes some .xip unpacking innovations thanks to theuni.

Thanks for working on this. This is great news. Maybe can manage to do it myself this time without a Mac :+1:

laanwj commented on contrib/macdeploy/README.md:
````diff
@@ -14,44 +14,45 @@ When complete, it will have produced `Bitcoin-Qt.dmg`.
+./contrib/macdeploy/gen_sdk_package.sh '/path/to/Xcode.app'
 ```
 
+### Historial macOS SDK Extraction Notes
````
I'm not sure keeping instructions around for previous SDKs is that useful. Only for the release that is used for building. Older information can always be found in git.

dongcarl replied:
Fixed!

dongcarl committed:
contrib: macdeploy: Remove historical extraction notes

theuni commented on contrib/macdeploy/README.md:
```diff
@@ -14,55 +14,41 @@ When complete, it will have produced `Bitcoin-Qt.dmg`.
+### Step 2: Generating `MacOSX10.14.sdk.tar.gz` from `Xcode.app`
+
+To generate `MacOSX10.14.sdk.tar.gz`, run the script
+[`gen_sdk_package.sh`](./gen_sdk_package.sh) with the XCODEDIR environment
```
"with the XCODEDIR environment variable pointing to...",
or
"./contrib/macdeploy/gen_sdk_package.sh '/path/to/Xcode.app'"

Which is it? :)

theuni replied:
gen-sdk.sh != gen-sdk.sh
(below as well)

theuni commented on contrib/macdeploy/gen-sdk.sh:
```diff
@@ -0,0 +1,121 @@
+die() { cat 1>&2 ; exit 1; }
+
+# Defaults to our current SDK's name, but overridable
+SDKNAME="${SDKNAME:-MacOSX10.14.sdk}"
```
Can we name this something to indicate that we've included the headers as well?

theuni reviewed (commented):
Thanks for doing this. Testing now.

theuni commented:
I'm seeing a Qt build failure due to some symlink/hardlink issue:
```
kernel/qdnslookup_unix.cpp:52:10: fatal error: 'arpa/nameser.h' file not found
#include <arpa/nameser.h>
         ^~~~~~~~~~~~~~~~
1 error generated.
Makefile:12337: recipe for target '.obj/qdnslookup_unix.o' failed
```
The link is broken hardlink in the generated sdk:
```
 $ ls -al MacOSX10.14.sdk/usr/include/arpa/nameser.h
lrwxrwxrwx 1 cory cory 26 Jun 12 14:28 MacOSX10.14.sdk/usr/include/arpa/nameser.h -> MacOSX10.14.sdk./nameser.h
```
Whereas in the original SDK it's a working symlink:
```
 $ ls -al MacOSX10.14.sdk/usr/include/arpa/nameser.h
lrwxrwxrwx 1 cory cory 12 Jun 12 14:28 MacOSX10.14.sdk/usr/include/arpa/nameser.h -> ../nameser.h
```

fanquake reviewed (commented):
Concept ACK.

Cool. Looks like Travis is happy now. Can you squash the fixup commits? 

I ran some gitian builds locally to compare the binaries produced by master (195822f1e05e2f36002c906667d4c639663f23b5) and master + this diff:
```diff
diff --git a/contrib/gitian-descriptors/gitian-osx.yml b/contrib/gitian-descriptors/gitian-osx.yml
index bbae7201e..936d097d2 100644
--- a/contrib/gitian-descriptors/gitian-osx.yml
+++ b/contrib/gitian-descriptors/gitian-osx.yml
@@ -32,7 +32,7 @@ remotes:
 - "url": "https://github.com/bitcoin/bitcoin.git"
   "dir": "bitcoin"
 files:
-- "MacOSX10.14.sdk.tar.gz"
+- "MacOSX10.14-with-libcxx-headers.sdk.tar.gz"
 script: |
   set -e -o pipefail
 
@@ -90,7 +90,7 @@ script: |
   BASEPREFIX="${PWD}/depends"
 
   mkdir -p ${BASEPREFIX}/SDKs
-  tar -C ${BASEPREFIX}/SDKs -xf ${BUILD_DIR}/MacOSX10.14.sdk.tar.gz
+  tar -C ${BASEPREFIX}/SDKs -xf ${BUILD_DIR}/MacOSX10.14-with-libcxx-headers.sdk.tar.gz
 
   # Build dependencies for each host
   for i in $HOSTS; do
diff --git a/depends/hosts/darwin.mk b/depends/hosts/darwin.mk
index 82e086a32..70baeae90 100644
--- a/depends/hosts/darwin.mk
+++ b/depends/hosts/darwin.mk
@@ -1,6 +1,6 @@
 OSX_MIN_VERSION=10.12
 OSX_SDK_VERSION=10.14
-OSX_SDK=$(SDK_PATH)/MacOSX$(OSX_SDK_VERSION).sdk
+OSX_SDK=$(SDK_PATH)/MacOSX$(OSX_SDK_VERSION)-with-libcxx-headers.sdk
 darwin_CC=clang -target $(host) -mmacosx-version-min=$(OSX_MIN_VERSION) --sysroot $(OSX_SDK)
 darwin_CXX=clang++ -target $(host) -mmacosx-version-min=$(OSX_MIN_VERSION) --sysroot $(OSX_SDK) -stdlib=libc++
 
diff --git a/depends/packages/native_cctools.mk b/depends/packages/native_cctools.mk
index 4195230b4..19fcf54f9 100644
--- a/depends/packages/native_cctools.mk
+++ b/depends/packages/native_cctools.mk
@@ -73,6 +73,5 @@ define $(package)_stage_cmds
   cp lib/libLTO.so $($(package)_staging_prefix_dir)/lib/ && \
   cp -rf lib/clang/$($(package)_clang_version)/include/* $($(package)_staging_prefix_dir)/lib/clang/$($(package)_clang_version)/include/ && \
   cp bin/llvm-dsymutil $($(package)_staging_prefix_dir)/bin/$(host)-dsymutil && \
-  if `test -d include/c++/`; then cp -rf include/c++/ $($(package)_staging_prefix_dir)/include/; fi && \
   if `test -d lib/c++/`; then cp -rf lib/c++/ $($(package)_staging_prefix_dir)/lib/; fi
```

The binaries produced by the "new" SDK are all slightly larger (except for `bitcoin-wallet`) than the binaries currently produced by master:

| Commit | `bitcoind` | `bitcoin-cli` | `bitcoin-qt` | `bitcoin-tx` | `bitcoin-wallet` | `test_bitcoin` |
| -------- | -------- | ---------- | --------- | ---------- | -------------- | ------------ | 
| Master | 9084960 | 399464 | 24920412 | 1042272 | 2721068 | 15010424 |
| This PR | 9135728 | 407680 | 24958868 | 1050504 | 2717020 | 15021368 |

I did a quick comparison of `bitcoind`, there is a single symbols difference:
```bash
--- 19.txt
+++ ea.txt
@@ -26,14 +26,15 @@
 U std::bad_exception::what() const
 U std::runtime_error::what() const
 U std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::find(char, unsigned long) const
 U std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::rfind(char, unsigned long) const
 U std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare(char const*) const
 U std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare(unsigned long, unsigned long, char const*) const
 U std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare(unsigned long, unsigned long, char const*, unsigned long) const
+T std::__1::enable_if<__can_be_converted_to_string_view<char, std::__1::char_traits<char>, std::__1::basic_string_view<char, std::__1::char_traits<char> > >::value, int>::type std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare<std::__1::basic_string_view<char, std::__1::char_traits<char> > >(std::__1::basic_string_view<char, std::__1::char_traits<char> > const&) const
 U std::__1::__shared_weak_count::__get_deleter(std::type_info const&) const
 U std::__1::__vector_base_common<true>::__throw_length_error() const
 U std::__1::__vector_base_common<true>::__throw_out_of_range() const
```

There is also one less initializer function called in the new `bitcoind`, however I haven't looked into that yet.

dongcarl commented:
Squashed in the fixups!

Sjors reviewed (commented):
Concept ACK. I did some testing with 1944c5ff798d149422bf6ac7deb999789eef64d2 on macOS and Ubuntu. Code looks reasonable, but mostly above my head.

I found `extract_xcode.py` to be deterministic, but `gen-sdk.sh` produces a different checksum depending on the OS.

theuni committed:
depends: bump MacOS toolchain

clang   6.0.1  -> 8.0.0
cctools 921    -> 949.0.1
ld64    409.12 -> 530

theuni committed:
macos: Bump to xcode 11.3.1 and 10.15 SDK

This gets us a newer SDK with c++17 support and retains 10.12
back-compat.

Co-authored-by: Carl Dong <contact@carldong.me>

dongcarl commented on contrib/macdeploy/README.md:
```diff
@@ -14,55 +14,42 @@ When complete, it will have produced `Bitcoin-Qt.dmg`.
-find Xcode.app -type d -name MacOSX.sdk -exec sh -c 'tar --transform="s/MacOSX.sdk/MacOSX10.14.sdk/" -c -C$(dirname {}) MacOSX.sdk/ | gzip -9n > MacOSX10.14.sdk.tar.gz' \;
+# Unpack Xcode_10.2.1.xip and place the resulting Xcode.app in your current
+# working directory
+python3 apple-sdk-tools/extract_xcode.py -f Xcode_10.2.1.xip | cpio -d -i
```
Nice!

dongcarl commented on contrib/macdeploy/README.md:
```diff
@@ -14,55 +14,42 @@ When complete, it will have produced `Bitcoin-Qt.dmg`.
-contrib/macdeploy/extract-osx-sdk.sh
-rm -rf 5.hfs MacOSX10.11.sdk
+# Generate a MacOSX10.14-with-libcxx-headers.sdk.tar.gz from the supplied Xcode.app
+./contrib/macdeploy/gen-sdk.sh '/path/to/Xcode.app'
```
That's somewhat expected, as bsdtar and gnutar (and even different versions of gnutar) might lay out the archive in different ways that all conform to standard. :relaxed:

Sjors replied:
If _you_ can live with a reproducibility issue, it must be fine :-)

theuni commented:
I discussed this today with @fanquake, and we realized that after this change, we'll also need to bump our SDK in the near-future so that we can begin testing with c++17. And when we do that, it's also a good time to bump our clang/cctools to match.

I hacked that together and tested quickly, and all seems to be working as expected:
https://github.com/theuni/bitcoin/tree/19240

At the risk of slowing down this PR, we could potentially integrate the rest of the toolchain bump in here as well. The upside of that would be that builders only have to download and generate a new sdk tarball once, whereas if we do this in stages, they'll have do it twice.

But then again, Carl has been maintaining these changes for quite a while now and is probably tired of waiting.

@dongcarl Your call :)

theuni commented on depends/packages/native_cctools.mk:
```diff
@@ -73,6 +73,5 @@ define $(package)_stage_cmds
   cp -rf lib/clang/$($(package)_clang_version)/include/* $($(package)_staging_prefix_dir)/lib/clang/$($(package)_clang_version)/include/ && \
   cp bin/llvm-dsymutil $($(package)_staging_prefix_dir)/bin/$(host)-dsymutil && \
-  if `test -d include/c++/`; then cp -rf include/c++/ $($(package)_staging_prefix_dir)/include/; fi && \
   if `test -d lib/c++/`; then cp -rf lib/c++/ $($(package)_staging_prefix_dir)/lib/; fi
```
This can be deleted now too, we know where to expect the c++ libs now.

Sjors commented:
Let's merge this first. As long as the other stuff gets in before 0.21 most people won't be impacted anyway.

dongcarl commented:
Thought a bit about what's needed for the steps that are required before 0.21, and it seems like I'll have to redo the entire `gen-sdk.sh` in python for the `tarfile` and `plistlib` modules, so it probably doesn't make sense to merge this in the meantime. Will push up a new version when I have it.

Many thanks to laanwj, theuni, fanquake, and especially Sjors for reviewing.

theuni commented:
> Let's merge this first.

Sounds good.

My only remaining ask for this PR is something @dongcarl and I discussed yesterday: Adding the xcode version string to the (already lengthy!) sdk tarball/dirname. This is to compensate for the fact that Apple updates the SDK in new XCode versions, so unfortunately "MacOSX10.xy.sdk" alone doesn't mean much. Adding this string will prevent future clashes on our end.

dongcarl committed:
Adapt rest of tooling to new SDK naming scheme

theuni commented:
@dongcarl Whoa, that was quick!

I haven't really reviewed the new script, but it worked great. Built locally, looks good. I went ahead and uploaded the resulting tarball for Travis so we can see where we stand.

Now that this includes a toolchain bump, mind updating the PR title/description? I think this is ready for in-depth review now.

Sjors commented on contrib/macdeploy/gen-sdk:
```diff
@@ -0,0 +1,94 @@
+    args = parser.parse_args()
+
+    xcode_app = pathlib.Path(args.xcode_app[0]).resolve()
+    assert xcode_app.is_dir(), f"The supplied Xcode.app path '{xcode_app}' either does not exist or is not a directory"
```
This is not Python 3.5.6 syntax (try with `pyenv`)

dongcarl replied:
Shame :frowning_face:, I like f-strings! Will redo with `.format`.

dongcarl replied:
Done and tested with Python 3.5.9!

Sjors reviewed (approved):
tACK 538bb0b (some minor issues for followup or if you need to touch things)

Gitian build:
```
019e21e451bdba8b803431ddf3be2bc8b2213da02d0163a23b2eda77a6ecb25e  bitcoin-538bb0bbad8d-osx-unsigned.dmg
fc5ea07e47b37adc24513225edc93ddc6a8f6867ae0f24049a560e5a2dd0883d  bitcoin-538bb0bbad8d-osx-unsigned.tar.gz
714bc66d4d7f0131dedd50b5cdfb6da4f1161da91bc2bbeda19f81787ff5756d  bitcoin-538bb0bbad8d-osx64.tar.gz
4958ef9e024dcf41044c0297b278e04b54e422710c6ad813c9bb5b162bd25148  src/bitcoin-538bb0bbad8d.tar.gz
4f5d9ba4ad9e0828cdb7e0c85587d51faf7850be5b8fc68d1b00fe3ccdf7efb3  bitcoin-core-osx-0.21-res.yml
```

Lightly tested DMG on macOS 10.15.5

fanquake commented on contrib/macdeploy/README.md:
```diff
@@ -14,55 +14,44 @@ When complete, it will have produced `Bitcoin-Qt.dmg`.
+Our current macOS SDK
+(`Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers.tar.gz`) can be
+extracted from
+[Xcode_11.3.1.xip](https://download.developer.apple.com/Developer_Tools/Xcode_11.3.1/Xcode_11.3.1.xip).
```
>  Also, maybe use 11.5?

NACK. 11.3.1 was picked specifically. It's what is referenced throughout the docs, build  system and  in the tarball name. There's also no guarantee that you'll generate the same tarball between various versions of Xcode.

Sjors replied:
What do you mean by "picked specifically"? It's introduced in this PR. The only other place I see it is on the Travis macOS native machine, but the choice there doesn't matter.

theuni replied:
Sorry for not explaining in more detail in the commit message.

11.3.1 was chosen as the last Xcode version which uses (Apple's version of) clang 8.x.x, for which we have a matching cctools and ld64. Starting with 11.4, they bumped to clang 9.x.x. Once cctools-port has a working matching version for clang 9, we can bump to a newer Xcode version.

dongcarl replied:
A thought: to prevent future confusion, we should follow up this PR with documentation on the reasoning behind how our macOS toolchain is constructed.

Sjors replied:
More info would be a good followup. I'm surprised Apple would bump clang 8 to 9 in a minor Xcode version update. But actually that could make sense, because they want to give developers access to SDK's for their next generation(s) OS, without forcing devs to upgrade their production machine to a beta/recent macOS. According to [wikipedia](https://en.wikipedia.org/wiki/Xcode#11.x_series) Xcode 11.3.1 requires macOs 10.14.4, whereas Xcode 11.4 bumps that requirement to macOS 10.15.2. The specific choice of clang 8.x.x makes sense.

Sjors commented on contrib/macdeploy/gen-sdk:
```diff
@@ -0,0 +1,94 @@
+        sdk_build_id = pl['ProductBuildVersion']
+    print("Found MacOSX SDK (version: {version}, build id: {build_id})".format(version=sdk_version, build_id=sdk_build_id))
+
+    out_name = "Xcode-{version}-{build_id}-extracted-SDK-with-libcxx-headers".format(version=sdk_version, build_id=sdk_build_id)
```
This should use `{xcode_version}`, not `{version}`; it broke when you changed the formatting.

dongcarl replied:
Good catch! Fixed.

Sjors reviewed (changes requested):
4a7932285ca7fc9a8e91af58567106b2ff04519d is almost there...

Sjors commented:
re-tACK cbdaa3a179c5011b651cb4c9cbcef482816d822d

Gitian build changed, presumable because you rebased (from a79bca2f1fb25f433d6e100a31a3acfde2656ce1 to febe5823b4ae0de5cd0e6da3f69555acd0724267)?

```
2f13a3069c1608f0e3c5ab0fa561e22273165acf710c8cc2fe399790b4fd56de  bitcoin-cbdaa3a179c5-osx-unsigned.dmg
fb00a5c1306c33fc54e3aafde51cdc12ab890785eda253717b3924529a6d2366  bitcoin-cbdaa3a179c5-osx-unsigned.tar.gz
4ea413de539ee7e972a5f4d60108016729d7cb92e6d15d5a2747b083fae87140  bitcoin-cbdaa3a179c5-osx64.tar.gz
1173147fd75c59bd44724b489280ea7895b61946a3c91f410d98b5770aa6250a  src/bitcoin-cbdaa3a179c5.tar.gz
dae1d7de69df0f8deb72a0138c74fcb32a4d70a4358e99b78ae2fa362834e4bb  bitcoin-core-osx-0.21-res.yml
```

dongcarl commented:
I can reproduce Sjor's results!

```
- out_manifest: |
    2f13a3069c1608f0e3c5ab0fa561e22273165acf710c8cc2fe399790b4fd56de  bitcoin-cbdaa3a179c5-osx-unsigned.dmg
    fb00a5c1306c33fc54e3aafde51cdc12ab890785eda253717b3924529a6d2366  bitcoin-cbdaa3a179c5-osx-unsigned.tar.gz
    4ea413de539ee7e972a5f4d60108016729d7cb92e6d15d5a2747b083fae87140  bitcoin-cbdaa3a179c5-osx64.tar.gz
    1173147fd75c59bd44724b489280ea7895b61946a3c91f410d98b5770aa6250a  src/bitcoin-cbdaa3a179c5.tar.gz
```

MarcoFalke commented:
Let me know when this is ready so I can put the backdoored binary on DrahtBot :)

MarcoFalke commented:
Before merge, the OP should be edited to remove the historic description, which is accessible through the GitHub history drop-down

MarcoFalke commented:
the ci failure can be ignored: https://github.com/bitcoin/bitcoin/issues/19329#issuecomment-646979700

MarcoFalke commented:
Concept ACK cbdaa3a179c5011b651cb4c9cbcef482816d822d (I read the OP and don't disagree)

dongcarl commented:
@MarcoFalke theuni said he'll run thru it once more on Monday, should be pretty safe after that!

fanquake commented on depends/packages/native_cctools.mk:
```diff
@@ -1,14 +1,14 @@
 package=native_cctools
-$(package)_version=3764b223c011574971ee3ae09ce968ba5dc2f00f
+$(package)_version=45fe81aceccb8f8b23f1d0f0ecfd13e29eecae98
```
```suggestion
$(package)_version=4da2f3b485bcf4cef526f30c0b8c0bcda99cdbb4
```

dongcarl replied:
Done!

theuni replied:
Nice spot, thanks.

fanquake reviewed (changes requested):
Looks good. Seems to align with what me and @theuni discussed. Tested `extract_xcode.py`. However, I noticed while building depends that the version of LD64 being passed around didn't look correct `-DLD64_VERSION_NUM=512.4`. It should be `530`. Looks like upstream didn't update it when doing the last LD64 update, however they have noticed and fixed it, so we should move to cctools @ https://github.com/tpoechtrager/cctools-port/commit/4da2f3b485bcf4cef526f30c0b8c0bcda99cdbb4, which is in the `949.0.1-ld64-530` branch.

dongcarl commented:
2418f739f75824d6689369f326b960cec254cf56 includes fanquake's recommendation [here](https://github.com/bitcoin/bitcoin/pull/19240#discussion_r443187584), and a modification that makes the history more readable (some leftover mistake from my squashing).

Should be ready to merge once theuni takes a last look today.

theuni committed:
darwin: pass mlinker-version so that clang enables new features

Without this clang fails to add any newly-added linker features.

Removing this in ca5055a5aa07aba81a87cf12f6f0526a63c423b5 was likely a
regression.

See https://github.com/bitcoin/bitcoin/pull/19240#issuecomment-647764049
for more discussion.

theuni commented:
Actually, after digging through @fanquake's recommendation some more, it looks like the LD64_VERSION_NUM does **not** end up being passed through correctly. We can poke at clang's source to see what it does with that value. One easy-to-demonstrate case can be seen [here](https://github.com/llvm/llvm-project/blob/master/clang/lib/Driver/ToolChains/Darwin.cpp#L217)
```c++
  // Newer linkers support -demangle. Pass it if supported and not disabled by
  // the user.
  if (Version[0] >= 100 && !Args.hasArg(options::OPT_Z_Xlinker__no_demangle))
    CmdArgs.push_back("-demangle");
```

Linking a test with our depends-generated toolchain:
> /home/cory/dev/bitcoin2/depends/x86_64-apple-darwin16/native/bin/clang++ -target x86_64-apple-darwin16 -mmacosx-version-min=10.12 --sysroot /home/cory/dev/bitcoin2/depends/SDKs/Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers -stdlib=libc++ -std=c++11 test.cpp -o testing -v

Shows the link-line:
> "/home/cory/dev/bitcoin2/depends/x86_64-apple-darwin16/native/bin/x86_64-apple-darwin16-ld" -dynamic -arch x86_64 -macosx_version_min 10.12.0 -syslibroot /home/cory/dev/bitcoin2/depends/SDKs/Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers -o testing /tmp/test-385d92.o -lc++ -lSystem

However, a test with -mlinker-version=100:

> /home/cory/dev/bitcoin2/depends/x86_64-apple-darwin16/native/bin/clang++ -target x86_64-apple-darwin16 -mmacosx-version-min=10.12 --sysroot /home/cory/dev/bitcoin2/depends/SDKs/Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers -stdlib=libc++ -std=c++11 test.cpp -o testing -v -mlinker-version=100


Yields:
> "/home/cory/dev/bitcoin2/depends/x86_64-apple-darwin16/native/bin/x86_64-apple-darwin16-ld" **-demangle** -dynamic -arch x86_64 -macosx_version_min 10.12.0 -syslibroot /home/cory/dev/bitcoin2/depends/SDKs/Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers -o testing /tmp/test-a0732e.o -lc++ -lSystem

Note that "-demangle" was added as-expected only when -mlinker-version was passed. Without that option, it seems to assume the lowest possible value. So I guess this is a regression from all the way back in ca5055a5aa07aba81a87cf12f6f0526a63c423b5 :(

I suggest we take https://github.com/theuni/bitcoin/commit/718180005fb031713445498d89cc0565e5a6f6ec as part of this PR as well. Then maybe we can send @fanquake down the rabbit hole to see why clang can't find this on its own (or if it's even supposed to).

dongcarl commented:
cherry-picked 718180005fb031713445498d89cc0565e5a6f6ec into this PR so that we only nuke depends cache once. @theuni anything else you want to address?

theuni reviewed (approved):
ACK adf543d7144a44f8cd28a090a21c4e2a606862da.

Thanks, everyone for sticking with this. The scope ended up expanding a few times, but I'm really happy to get this all in at once.

Big thanks to @Sjors as well for testing along the way.

dongcarl commented:
Gitian build succeeded locally:
```yml
- out_manifest: |
    5eb9795549bc1f3a91cfa4b8901db357343d6386a68c15298009546ed28f3f74  bitcoin-adf543d7144a-osx-unsigned.dmg
    8f03081ef08035ee91c5fa10e1ddb86842cb318dd62ea923e7346c6f4306cd96  bitcoin-adf543d7144a-osx-unsigned.tar.gz
    f065ab8d37a473c180bb92f7613de03ba3a1b5f604f0aaa2cd4dd8d9953a9e7f  bitcoin-adf543d7144a-osx64.tar.gz
    ffc2e29782439497a2a27de16996d6b3de5225feb06e6f63ece3680557470e37  src/bitcoin-adf543d7144a.tar.gz
```

fanquake reviewed (approved):
ACK adf543d7144a44f8cd28a090a21c4e2a606862da - I'll take a look at the linker issue.

Sjors commented:
Post merge: I get the same gitian result as @dongcarl found.

MarcoFalke commented:
Is the `Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers.tar.gz` deterministic and where can I get it without an apple account?

MarcoFalke commented:
@jonasschnelli https://bitcoin.jonasschnelli.ch/gitian/#nighly also needs the new tar.gz

theuni commented:
> Is the `Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers.tar.gz` deterministic and where can I get it without an apple account?

The .tar.gz is not deterministic for the usual tarball reasons, but its contents are.

I'm afraid you'll need an apple account to grab Xcode.

Sjors commented:
@MarcoFalke you can compare your results with mine: https://github.com/bitcoin/bitcoin/pull/19240#discussion_r442705416

MarcoFalke commented:
Could someone please upload a mac build.log for me to compare? I get a mismatch
